### ISSUE TITLE
UTM app does not map certain keys for the non-english keyboards.

### ISSUE THREAD

**Issue by ozbrk** [utmapp/UTM#3175]
**Describe the issue**
UTM app does not map certain keys for the non-english keyboards.

Keys that fail

option + q supposed to print @ but it printing q 
" and é are at the same key but it is unresponsive
> and < are at the same key but it prints "

There seems to be no other issue. Keyboard language is Turkish / TR 

**Configuration**
* UTM Version: Version 2.2.4 (36)
* OS Version:  Big Sur
* Intel or Apple Silicon?:  Apple Silicon M1

**Crash log**
App does not crash 

**Debug log**
Running:  -L /Applications/UTM.app/Contents/Resources/qemu -S -qmp tcp:127.0.0.1:4001,server,nowait -nodefaults -vga none -spice "unix=on,addr=/Users/ozberkakbas/Library/Group Containers/WDNLXAD4W8.com.utmapp.UTM/3B2EE660-5AB5-4A89-AE8B-6BF7397230BC.spice,disable-ticketing=on,image-compression=off,playback-compression=off,streaming-video=off,gl=on" -device virtio-ramfb-gl -cpu cortex-a72 -smp cpus=8,sockets=1,cores=8,threads=1 -machine virt-6.1,highmem=off -accel hvf -accel tcg,tb-size=1024 -drive if=pflash,format=raw,unit=0,file=/Applications/UTM.app/Contents/Resources/qemu/edk2-aarch64-code.fd,readonly=on -drive if=pflash,format=raw,unit=1,file=/Users/ozberkakbas/Library/Containers/com.utmapp.UTM/Data/Documents/Ubuntu.utm/Images/efi_vars.fd -boot menu=on -m 4096 -device intel-hda -device hda-duplex -name Ubuntu -device qemu-xhci,id=usb-bus -device usb-tablet,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -device usb-kbd,bus=usb-bus.0 -device ich9-usb-ehci1,id=usb-controller-0 -device ich9-usb-uhci1,masterbus=usb-controller-0.0,firstport=0,multifunction=on -device ich9-usb-uhci2,masterbus=usb-controller-0.0,firstport=2,multifunction=on -device ich9-usb-uhci3,masterbus=usb-controller-0.0,firstport=4,multifunction=on -chardev spicevmc,name=usbredir,id=usbredirchardev0 -device usb-redir,chardev=usbredirchardev0,id=usbredirdev0,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev1 -device usb-redir,chardev=usbredirchardev1,id=usbredirdev1,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev2 -device usb-redir,chardev=usbredirchardev2,id=usbredirdev2,bus=usb-controller-0.0 -device nvme,drive=drive0,serial=drive0,bootindex=0 -drive if=none,media=disk,id=drive0,file=/Users/ozberkakbas/Library/Containers/com.utmapp.UTM/Data/Documents/Ubuntu.utm/Images/disk-0.qcow2,cache=writethrough -device usb-storage,drive=drive1,removable=true,bootindex=1 -drive if=none,media=cdrom,id=drive1 -device virtio-net-pci,mac=46:52:FC:E9:F2:BA,netdev=net0 -netdev vmnet-macos,mode=shared,id=net0 -device virtio-serial -device virtserialport,chardev=vdagent,name=com.redhat.spice.0 -chardev spicevmc,id=vdagent,debug=0,name=vdagent -uuid 3B2EE660-5AB5-4A89-AE8B-6BF7397230BC -rtc base=localtime
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: Started vmnet interface with configuration:
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: MTU:              1500
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: Max packet size:  1514
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: MAC:              3e:14:6c:42:0f:84
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: DHCP IPv4 start:  192.168.64.1
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: DHCP IPv4 end:    192.168.64.254
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: IPv4 subnet mask: 255.255.255.0
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: UUID:             1FABB9AB-E693-4762-8192-7941489DDAC0
qemu-aarch64-softmmu: warning: Spice: playback:0 (0x139897120): setsockopt failed, Operation not supported on socket
qemu-aarch64-softmmu: warning: Spice: record:0 (0x1398971d0): setsockopt failed, Operation not supported on socket
gl_version 30 - es profile enabled
WARNING: running without ARB/KHR robustness in place may crash
GLSL feature level 130

**Upload VM**
N/A

**Comment by osy** [comment]
@ozbrk can you test v2.3.0? There’s been some recent changes on how characters are mapped.

**Comment by ozbrk** [comment]
@conath it is set to the Turkish. Maybe key codes on macOS for Turkish keyboard are different than the international? But in that case the result would have been different keystrokes rather than no-response from one key. 

option + q is the @ usually. But it prints nothing instead it generates a "bip" sound and prints q 
" key is next to tab and unresponsive
[<] and [>] are next to left shift. It prints ". With option it must be | but it generates nothing.

**Comment by conath** [comment]
Is both macOS AND the VM’s operating system set to the same Turkish layout?

**Comment by conath** [comment]
Thanks for the added info!

**Comment by ozbrk** [comment]
I'm volunteering to help. The thing is there is no other virtual machine solution for the M1 MacBook out there and I'm willing to help.

**Comment by ozbrk** [comment]
@conath I tried. On macOS I have used US International and on vm I have used TR 
Nothing changed yet I get the Turkish keys on the VM

**Comment by ozbrk** [comment]
Yeah. But I couldn't get the @ and other key combinations. Other TR keys are worked. (Just wanted to made this clear)

**Comment by ozbrk** [comment]
I might have some useful information. Not from the UTM but another application called Any Desk. Now the story is, when I connect my bluetooth keyboard (which is macOS compatible) to any desk and started a remote session to my sandbox machine I saw a strange thing. 

On Windows situation was like this:

[" , é] keys are replaced with  [ < , > , | ] key.

This might not even be related but I think we need to know more how Mac OS actually maps the keyboard on itself.

**Comment by sanberkzulfikar** [comment]
I am also experiencing this problem with the same keyboard layout (Turkish Q) with M1 MacBook Air 16GB RAM 7Core GPU configuration.

I am using the latest UTM (v.2.3.0) and it occurs both on Ubuntu LTS 20.4.3 and Windows 11 ARM Insider Preview.
For example, when I try to produce the "#" symbol, the virtual machines won't recognize the option key.

When I tried on my MacBook installing the same Ubuntu image file on VMware Fusion Tech Preview for ARM, the problem did not occur so it must be something about UTM.

**Comment by ozbrk** [comment]
Is there any other update on that matter? I wish I could really help on this but my knowledge on swift is really low. :(

**Comment by ozbrk** [comment]
" think we have to either find a way to get the raw hardware scancode and pass it directly to the VM or have some sort of secondary translation for region specific keyboards. The second case will be quite time consuming due to the number of keyboards that are out there." 

First question how can we extract the scan code 
Second question can we automatize the processs for example can we create a script that is able to extract the scan code from the keyboard layout for us.
Third question can we ask people to run the script and automatically localize the codes into the UTM

**Comment by osy** [comment]
Well step one is if you can recommend me a good way to test a virtual Turkish keyboard in ubuntu.

**Comment by ozbrk** [comment]
Plus;
xvkbd might not work because there is no aarch64 edition I think. 
https://forums.raspberrypi.com/viewtopic.php?t=235163  
in here there are alternatives suggested for it

**Comment by ozbrk** [comment]
Plus plus;

I found this on the internet while wandering about the issue 👍 
http://florence.sourceforge.net/english/alternatives.html 
Florence might work for you

**Comment by ozbrk** [comment]
It's fixed. Only issue is MacBook making an alert "beep" as I pressed. If you have any idea how to fix it we are good to go now. Thank you for all of your efforts my friend .

**Comment by simonhuwiler** [comment]
I got similar issues on a Swiss (de_ch) Keyboard. Its sending the wrong keys:

**Clipboard:**
Should: `Command L` + `C`. 
Is: `Control L` + `C`. 

**@:**
Should: `Option L` + `g`. 
Is: `Option R` + `g`. 

Based on the virtual keyboard, the Keys are mapped to following (host -> VM):
* Control L -> Control L
* Option L -> Meta L
* Command L -> ?? (not showing on the virtual keyboard, but switching window)
* Command R -> Super R
* Option R -> Level 3 S...
* Control R -> Control R

Host: MacBook Pro, M1, BigSure
UTM: 3.2.4
VM: Ubuntu 20.0.4
Keyboard-Layout on Ubuntu: `German (Switzerland, Macintosh)`

**Comment by conath** [comment]
@simonhuwiler I am not sure I understand the issue. Ubuntu uses different key combinations for copy, paste and '@' than macOS. 

AFAIK the host keyboard layout does not affect the keys sent to the guest.

**Comment by simonhuwiler** [comment]
I'm an idiot. Sorry for bothering.
